File: app/src/main/java/uk/ac/swansea/codenames/start_screen.java
package uk.ac.swansea.codenames;

import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.TextView;

public class start_screen extends AppCompatActivity {

    // TODO: Add an exit button.

    private androidx.gridlayout.widget.GridLayout quitBox;
    private ConstraintLayout constraintLayout;
    private Button playButton;
    private Button settingsButton;
    private Button yesButton;
    private Button noButton;
    private TextView quitText;
    private TextView startScreenTitle;
    private int defaultColour;
    private boolean exiting = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
        setContentView(R.layout.start_screen);

        quitBox = findViewById(R.id.quitBox);
        playButton = findViewById(R.id.playButton);
        settingsButton = findViewById(R.id.settingsButton);
        constraintLayout = findViewById(R.id.saveButton);
        yesButton = findViewById(R.id.yesButton);
        noButton = findViewById(R.id.noButton);
        quitText = findViewById(R.id.quitText);
        startScreenTitle = findViewById(R.id.startScreenTitle);

        userSettings.getInstance().Initialize(getApplicationContext());

        //This is how TTS will work. Pressing the button quickly performs it's default action. Holding it
        //with a return true in this listener, will cause it to run the code in here, and not run the code
        //for a regular click.
        //For actual TTS it won't print out the word, it'll obviously read it aloud
        playButton.setOnLongClickListener(v -> {
            System.out.println(playButton.getText().toString());
            return true;
        });

        updateColours();
    }

    @Override
    public void onBackPressed() {
        if (exiting) {
            quitBox.setVisibility(View.INVISIBLE);
            playButton.setVisibility(View.VISIBLE);
            settingsButton.setVisibility(View.VISIBLE);

            exiting = false;
        } else {
            quitBox.setVisibility(View.VISIBLE);
            playButton.setVisibility(View.INVISIBLE);
            settingsButton.setVisibility(View.INVISIBLE);

            exiting = true;
        }
    }

    public void toMainMenu(View view) {
        Intent i = new Intent(this, main_menu.class);
        startActivity(i);
    }

    public void toSettings(View view) {
        Intent i = new Intent(this, settings.class);
        i.putExtra("from", "start_screen");
        startActivity(i);
    }

    public void quitGame(View view) {
        this.finishAffinity();
        System.exit(0);
    }

    public void noQuit(View view) {
        quitBox.setVisibility(View.INVISIBLE);
        playButton.setVisibility(View.VISIBLE);
        settingsButton.setVisibility(View.VISIBLE);

        exiting = false;
    }

    public void updateColours() {
        if (userSettings.getInstance().getPreference(userSettings.getInstance().APPLICATION_BACKGROUND).equals("")) {
            defaultColour = userSettings.getInstance().APPLICATION_BACKGROUND_DEFAULT;
        } else {
            defaultColour = Integer.parseInt(userSettings.getInstance().getPreference(userSettings.getInstance().APPLICATION_BACKGROUND));
        }

        constraintLayout.setBackgroundColor(defaultColour);
        quitBox.setBackgroundColor(defaultColour);

        if (userSettings.getInstance().getPreference(userSettings.getInstance().MENU_BUTTONS).equals("")) {
            defaultColour = userSettings.getInstance().MENU_BUTTONS_DEFAULT;
        } else {
            defaultColour = Integer.parseInt(userSettings.getInstance().getPreference(userSettings.getInstance().MENU_BUTTONS));
        }

        playButton.setBackgroundColor(defaultColour);
        settingsButton.setBackgroundColor(defaultColour);
        yesButton.setBackgroundColor(defaultColour);
        noButton.setBackgroundColor(defaultColour);

        if (userSettings.getInstance().getPreference(userSettings.getInstance().MENU_TEXT).equals("")) {
            defaultColour = userSettings.getInstance().MENU_TEXT_DEFAULT;
        } else {
            defaultColour = Integer.parseInt(userSettings.getInstance().getPreference(userSettings.getInstance().MENU_TEXT));
        }

        quitText.setTextColor(defaultColour);
        startScreenTitle.setTextColor(defaultColour);
        playButton.setTextColor(defaultColour);
        settingsButton.setTextColor(defaultColour);
    }
}

File: app/src/main/java/uk/ac/swansea/codenames/StartScreen.kt
package uk.ac.swansea.codenames

import android.content.Context
import android.content.Intent
import android.media.MediaPlayer
import android.os.Bundle
import android.speech.tts.TextToSpeech
import android.view.View
import android.widget.ImageView
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintLayout
import com.google.android.material.button.MaterialButton
import com.google.android.material.textview.MaterialTextView
import java.util.*
import kotlin.system.exitProcess

/**
 * Starting screen, this is what you will see when you launch the app for the 2nd and onwards time.
 * You can go to the settings or main menu screen.
 */
class StartScreen : AppCompatActivity(), TextToSpeech.OnInitListener {
    private var quitBox: ConstraintLayout? = null
    private var constraintLayout: ConstraintLayout? = null
    private var playButton: MaterialButton? = null
    private var settingsButton: MaterialButton? = null
    private var exitButton: MaterialButton? = null
    private var yesButton: MaterialButton? = null
    private var noButton: MaterialButton? = null
    private var quitText: MaterialTextView? = null
    private var startScreenTitle: MaterialTextView? = null
    private var playImage: ImageView? = null
    private var settingsImage: ImageView? = null
    private var applicationBackgroundColour = -10921639
    private var menuButtonsColour = -8164501
    private var menuTextColour = -1
    private var exiting = false
    private var textToSpeech: TextToSpeech? = null
    private var buttonClick: MediaPlayer? = null

    /**
     * Sets up the layout for the screen.
     * Adds references to each View, sets up listeners and calls the updateColours function.
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.start_screen)

        quitBox = findViewById(R.id.quitBox)
        playButton = findViewById(R.id.playButton)
        settingsButton = findViewById(R.id.settingsButton)
        exitButton = findViewById(R.id.exitButton)
        constraintLayout = findViewById(R.id.constraintLayout)
        yesButton = findViewById(R.id.yesButton)
        noButton = findViewById(R.id.noButton)
        quitText = findViewById(R.id.quitText)
        startScreenTitle = findViewById(R.id.startScreenTitle)
        playImage = findViewById(R.id.playImage)
        settingsImage = findViewById(R.id.settingsImage)

        buttonClick = MediaPlayer.create(this, R.raw.buttonclick)

        val preferences = this.getSharedPreferences("preferences", Context.MODE_PRIVATE)

        val soundFXVolume = preferences.getFloat("soundFXVolume", 0.5f)

        buttonClick?.setVolume(soundFXVolume, soundFXVolume)

        val intent = Intent(this, BackgroundMusicService::class.java)

        this.startService(intent)

        textToSpeech = TextToSpeech(this, this)

        playButton!!.setOnClickListener {
            textToSpeech?.stop()

            buttonClick?.start()

            val i = Intent(this, MainMenu::class.java)
            startActivity(i)
        }

        settingsButton!!.setOnClickListener {
            textToSpeech?.stop()

            buttonClick?.start()

            val i = Intent(this, Settings::class.java)
            i.putExtra("from", "start_screen")
            startActivity(i)
        }

        exitButton!!.setOnClickListener {
            textToSpeech?.stop()

            buttonClick?.start()

            onBackPressed()
        }

        yesButton!!.setOnClickListener {
            textToSpeech?.stop()

            buttonClick?.start()

            //Closes the app.
            finishAffinity()
            exitProcess(0)
        }

        noButton!!.setOnClickListener {
            textToSpeech?.stop()

            buttonClick?.start()

            quitBox!!.visibility = View.INVISIBLE
            playButton!!.visibility = View.VISIBLE
            settingsButton!!.visibility = View.VISIBLE
            exitButton!!.visibility = View.VISIBLE
            playImage!!.visibility = View.VISIBLE
            settingsImage!!.visibility = View.VISIBLE
            startScreenTitle!!.visibility = View.VISIBLE
            exiting = false
        }

        playButton!!.setOnLongClickListener {
            speakOut(playButton?.text.toString())
            true
        }

        settingsButton!!.setOnLongClickListener {
            speakOut(settingsButton?.text.toString())
            true
        }

        exitButton!!.setOnLongClickListener {
            speakOut(exitButton?.text.toString())
            true
        }

        yesButton!!.setOnLongClickListener {
            speakOut(yesButton?.text.toString())
            true
        }

        noButton!!.setOnLongClickListener {
            speakOut(noButton?.text.toString())
            true
        }

        quitText!!.setOnLongClickListener {
            speakOut(quitText?.text.toString().replace("\n",""))
            true
        }

        startScreenTitle!!.setOnLongClickListener {
            speakOut(startScreenTitle?.text.toString())
            true
        }

        updateColours()
    }

    /**
     * Sets up the TextToSpeech engine.
     */
    override fun onInit(status: Int) {
        if (status == TextToSpeech.SUCCESS) {
            textToSpeech!!.language = Locale.UK
        }
    }

    /**
     * Opens a box giving the user the choice to close the app
     * completely or not.
     */
    override fun onBackPressed() {
        textToSpeech?.stop()

        if (exiting) {
            quitBox!!.visibility = View.INVISIBLE
            playButton!!.visibility = View.VISIBLE
            settingsButton!!.visibility = View.VISIBLE
            exitButton!!.visibility = View.VISIBLE
            playImage!!.visibility = View.VISIBLE
            settingsImage!!.visibility = View.VISIBLE
            startScreenTitle!!.visibility = View.VISIBLE
            exiting = false
        } else {
            speakOut(quitText?.text.toString().replace("\n",""))

            quitBox!!.visibility = View.VISIBLE
            playButton!!.visibility = View.INVISIBLE
            settingsButton!!.visibility = View.INVISIBLE
            exitButton!!.visibility = View.INVISIBLE
            playImage!!.visibility = View.INVISIBLE
            settingsImage!!.visibility = View.INVISIBLE
            startScreenTitle!!.visibility = View.INVISIBLE
            exiting = true
        }
    }

    /**
     * Reads out the given string using the TextToSpeech engine if it is enabled.
     */
    private fun speakOut(message : String) {
        val preferences = this.getSharedPreferences("preferences", Context.MODE_PRIVATE)

        if (preferences.getBoolean("textToSpeech", true)) {
            textToSpeech!!.speak(message, TextToSpeech.QUEUE_FLUSH, null, "")
        }
    }

    /**
     * Applies the chosen colour scheme to the layout.
     */
    fun updateColours() {
        val preferences = this.getSharedPreferences("preferences", Context.MODE_PRIVATE)

        applicationBackgroundColour = preferences.getInt("applicationBackground", -10921639)
        menuButtonsColour = preferences.getInt("menuButtons", -8164501)
        menuTextColour = preferences.getInt("menuText", -1)

        constraintLayout!!.setBackgroundColor(applicationBackgroundColour)
        quitBox!!.setBackgroundColor(applicationBackgroundColour)

        playButton!!.setBackgroundColor(menuButtonsColour)
        settingsButton!!.setBackgroundColor(menuButtonsColour)
        exitButton!!.setBackgroundColor(menuButtonsColour)
        yesButton!!.setBackgroundColor(menuButtonsColour)
        noButton!!.setBackgroundColor(menuButtonsColour)
        playImage!!.setColorFilter(menuButtonsColour)
        settingsImage!!.setColorFilter(menuButtonsColour)

        quitText!!.setTextColor(menuTextColour)
        startScreenTitle!!.setTextColor(menuTextColour)
        playButton!!.setTextColor(menuTextColour)
        settingsButton!!.setTextColor(menuTextColour)
        exitButton!!.setTextColor(menuTextColour)
    }
}